File: examples/hello-world/gabriel-greeting-rust/internal/server_test.rs
use crate::gen::rust::greeting::v1 as pb;
use holons::gen::holons::v1::any_value;
use holons::observability::{self, Field};
use std::collections::HashMap;
use tokio::net::TcpListener;
use tokio::sync::oneshot;
use tokio_stream::wrappers::TcpListenerStream;
use tonic::Request;

struct TestServer {
    client: pb::greeting_service_client::GreetingServiceClient<tonic::transport::Channel>,
    shutdown: Option<oneshot::Sender<()>>,
}

impl Drop for TestServer {
    fn drop(&mut self) {
        if let Some(shutdown) = self.shutdown.take() {
            let _ = shutdown.send(());
        }
    }
}

async fn start_server() -> TestServer {
    let listener = TcpListener::bind("127.0.0.1:0")
        .await
        .expect("should bind an ephemeral test port");
    let address = listener
        .local_addr()
        .expect("listener should have an address");
    let incoming = TcpListenerStream::new(listener);
    let (shutdown_tx, shutdown_rx) = oneshot::channel();

    tokio::spawn(async move {
        let result = tonic::transport::Server::builder()
            .add_service(pb::greeting_service_server::GreetingServiceServer::new(
                crate::internal::server::GreetingServer,
            ))
            .serve_with_incoming_shutdown(incoming, async {
                let _ = shutdown_rx.await;
            })
            .await;

        if let Err(error) = result {
            panic!("test server failed: {error}");
        }
    });

    let client =
        pb::greeting_service_client::GreetingServiceClient::connect(format!("http://{address}"))
            .await
            .expect("client should connect to the test server");

    TestServer {
        client,
        shutdown: Some(shutdown_tx),
    }
}

#[tokio::test]
async fn list_languages_returns_all_languages() {
    let mut server = start_server().await;

    let response = server
        .client
        .list_languages(pb::ListLanguagesRequest::default())
        .await
        .expect("ListLanguages should succeed")
        .into_inner();

    assert_eq!(response.languages.len(), 56);
}

#[tokio::test]
async fn list_languages_populates_required_fields() {
    let mut server = start_server().await;

    let response = server
        .client
        .list_languages(pb::ListLanguagesRequest::default())
        .await
        .expect("ListLanguages should succeed")
        .into_inner();

    for language in response.languages {
        assert!(!language.code.is_empty());
        assert!(!language.name.is_empty());
        assert!(!language.native.is_empty());
    }
}

#[tokio::test]
async fn say_hello_uses_requested_language() {
    let mut server = start_server().await;

    let response = server
        .client
        .say_hello(pb::SayHelloRequest {
            name: "Bob".to_string(),
            lang_code: "fr".to_string(),
        })
        .await
        .expect("SayHello should succeed")
        .into_inner();

    assert_eq!(response.greeting, "Bonjour Bob");
    assert_eq!(response.language, "French");
    assert_eq!(response.lang_code, "fr");
}

#[tokio::test]
async fn say_hello_uses_localized_default_name() {
    let mut server = start_server().await;

    let response = server
        .client
        .say_hello(pb::SayHelloRequest {
            name: String::new(),
            lang_code: "fr".to_string(),
        })
        .await
        .expect("SayHello should succeed")
        .into_inner();

    assert_eq!(response.greeting, "Bonjour Marie");
    assert_eq!(response.lang_code, "fr");
}

#[tokio::test]
async fn say_hello_falls_back_to_english() {
    let mut server = start_server().await;

    let response = server
        .client
        .say_hello(pb::SayHelloRequest {
            name: "Bob".to_string(),
            lang_code: "xx".to_string(),
        })
        .await
        .expect("SayHello should succeed")
        .into_inner();

    assert_eq!(response.greeting, "Hello Bob");
    assert_eq!(response.lang_code, "en");
}

#[tokio::test]
async fn say_hello_emits_otlp_shaped_log_record() {
    observability::reset();
    let mut env = HashMap::new();
    env.insert("OP_OBS".to_string(), "logs,metrics".to_string());
    let obs = observability::configure_from_env(
        observability::Config {
            slug: "gabriel-greeting-rust".to_string(),
            instance_uid: "greeting-test-uid".to_string(),
            ..observability::Config::default()
        },
        &env,
    )
    .unwrap();

    use pb::greeting_service_server::GreetingService;
    let response = crate::internal::server::GreetingServer
        .say_hello(Request::new(pb::SayHelloRequest {
            name: "Ana".to_string(),
            lang_code: "es".to_string(),
        }))
        .await
        .unwrap()
        .into_inner();
    assert_eq!(response.greeting, "Hola Ana");

    let record = obs
        .log_ring
        .as_ref()
        .unwrap()
        .drain()
        .into_iter()
        .find(|entry| entry.message == "Greeted Ana in Spanish (es)")
        .expect("missing greeting log");
    let wire = observability::to_proto_log_record(&record);
    assert_eq!(
        wire.body.as_ref().unwrap().value.as_ref(),
        Some(&any_value::Value::StringValue(
            "Greeted Ana in Spanish (es)".to_string()
        ))
    );
    assert_eq!(
        record.fields.get("transport"),
        Some(&Field::String("unknown".to_string()))
    );
    assert!(matches!(
        record.fields.get("duration_ns"),
        Some(Field::Int64(value)) if *value >= 0
    ));
    assert_eq!(
        observability::string_attribute(&wire.attributes, observability::ATTR_HOLONS_SLUG),
        "gabriel-greeting-rust"
    );
    assert_eq!(
        observability::string_attribute(&wire.attributes, observability::ATTR_SERVICE_NAME),
        "gabriel-greeting-rust"
    );
    assert_eq!(
        observability::string_attribute(&wire.attributes, observability::ATTR_HOLONS_INSTANCE_UID),
        "greeting-test-uid"
    );
    assert_eq!(
        observability::string_attribute(&wire.attributes, observability::ATTR_SERVICE_INSTANCE_ID),
        "greeting-test-uid"
    );

    observability::reset();
}

File: examples/hello-world/gabriel-greeting-go/internal/server_test.go
package internal_test

import (
	"context"
	"net"
	"strings"
	"testing"

	pb "gabriel-greeting-go/gen/go/greeting/v1"
	"gabriel-greeting-go/internal"

	v1 "github.com/organic-programming/go-holons/gen/go/holons/v1"
	"github.com/organic-programming/go-holons/pkg/observability"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/test/bufconn"
)

// startServer launches the GreetingService on an in-memory connection.
func startServer(t *testing.T) pb.GreetingServiceClient {
	t.Helper()
	lis := bufconn.Listen(1024 * 1024)
	s := grpc.NewServer()
	pb.RegisterGreetingServiceServer(s, &internal.Server{})

	go func() {
		if err := s.Serve(lis); err != nil {
			t.Logf("server exited: %v", err)
		}
	}()
	t.Cleanup(s.GracefulStop)

	conn, err := grpc.NewClient("passthrough:///bufconn",
		grpc.WithContextDialer(func(ctx context.Context, _ string) (net.Conn, error) {
			return lis.DialContext(ctx)
		}),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		t.Fatalf("grpc.NewClient: %v", err)
	}
	t.Cleanup(func() { conn.Close() })
	return pb.NewGreetingServiceClient(conn)
}

func TestListLanguages_ReturnsAll(t *testing.T) {
	client := startServer(t)
	resp, err := client.ListLanguages(context.Background(), &pb.ListLanguagesRequest{})
	if err != nil {
		t.Fatalf("ListLanguages: %v", err)
	}
	if len(resp.Languages) != 56 {
		t.Errorf("expected 56 languages, got %d", len(resp.Languages))
	}
}

func TestListLanguages_HasRequiredFields(t *testing.T) {
	client := startServer(t)
	resp, err := client.ListLanguages(context.Background(), &pb.ListLanguagesRequest{})
	if err != nil {
		t.Fatalf("ListLanguages: %v", err)
	}
	for _, lang := range resp.Languages {
		if lang.Code == "" {
			t.Error("found language with empty Code")
		}
		if lang.Name == "" {
			t.Errorf("language %q has empty Name", lang.Code)
		}
		if lang.Native == "" {
			t.Errorf("language %q has empty Native", lang.Code)
		}
	}
}

func TestSayHello_Nominal(t *testing.T) {
	client := startServer(t)
	resp, err := client.SayHello(context.Background(), &pb.SayHelloRequest{
		Name:     "Alice",
		LangCode: "fr",
	})
	if err != nil {
		t.Fatalf("SayHello: %v", err)
	}
	if resp.Greeting != "Bonjour Alice" {
		t.Errorf("expected 'Bonjour Alice', got %q", resp.Greeting)
	}
	if resp.Language != "French" {
		t.Errorf("expected language 'French', got %q", resp.Language)
	}
	if resp.LangCode != "fr" {
		t.Errorf("expected lang_code 'fr', got %q", resp.LangCode)
	}
}

func TestSayHello_EmptyName(t *testing.T) {
	client := startServer(t)
	resp, err := client.SayHello(context.Background(), &pb.SayHelloRequest{
		LangCode: "en",
	})
	if err != nil {
		t.Fatalf("SayHello: %v", err)
	}
	if resp.Greeting != "Hello Mary" {
		t.Errorf("expected 'Hello Mary', got %q", resp.Greeting)
	}
}

func TestSayHello_UnknownLanguageFallsBackToEnglish(t *testing.T) {
	client := startServer(t)
	resp, err := client.SayHello(context.Background(), &pb.SayHelloRequest{
		Name:     "Bob",
		LangCode: "xx",
	})
	if err != nil {
		t.Fatalf("SayHello: %v", err)
	}
	if resp.LangCode != "en" {
		t.Errorf("expected fallback to 'en', got %q", resp.LangCode)
	}
	if resp.Greeting != "Hello Bob" {
		t.Errorf("expected 'Hello Bob', got %q", resp.Greeting)
	}
}

func TestSayHello_EmitsObservabilitySignals(t *testing.T) {
	observability.Reset()
	t.Setenv("OP_OBS", "logs,metrics")
	obs := observability.Configure(observability.Config{Slug: "gabriel-greeting-go", InstanceUID: "greeting-test-uid"})
	t.Cleanup(func() {
		_ = obs.Close()
		observability.Reset()
	})

	resp, err := (&internal.Server{}).SayHello(context.Background(), &pb.SayHelloRequest{
		Name:     " Bob ",
		LangCode: "en",
	})
	if err != nil {
		t.Fatalf("SayHello: %v", err)
	}
	if resp.Greeting != "Hello Bob" {
		t.Fatalf("expected greeting 'Hello Bob', got %q", resp.Greeting)
	}

	snap := observability.Current().Registry().Snapshot()
	var foundCounter bool
	for _, sample := range snap.Counters {
		if sample.Name == "greeting_emitted_total" &&
			sample.Labels["lang_code"] == "en" &&
			sample.Labels["language"] == "English" &&
			sample.Labels["transport"] == "unknown" {
			foundCounter = true
			if sample.Value != 1 {
				t.Fatalf("greeting_emitted_total = %d, want 1", sample.Value)
			}
		}
	}
	if !foundCounter {
		t.Fatalf("missing greeting_emitted_total counter in %+v", snap.Counters)
	}

	var foundLog bool
	for _, entry := range observability.Current().LogRing().Drain() {
		record := observability.ToProtoLogRecord(entry)
		attrs := keyValues(record.GetAttributes())
		body := record.GetBody().GetStringValue()
		if strings.HasPrefix(body, "Greeted ") &&
			strings.HasSuffix(body, " (en)") &&
			attrs["lang_code"].GetStringValue() == "en" {
			foundLog = true
			if body != "Greeted Bob in English (en)" {
				t.Fatalf("unexpected greeting log message: %q", body)
			}
			if record.GetSeverityNumber() != v1.SeverityNumber_SEVERITY_NUMBER_INFO {
				t.Fatalf("severity_number = %v, want INFO", record.GetSeverityNumber())
			}
			if attrs[observability.AttrHolonsSlug].GetStringValue() != "gabriel-greeting-go" ||
				attrs[observability.AttrHolonsInstanceUID].GetStringValue() != "greeting-test-uid" ||
				attrs[observability.AttrServiceName].GetStringValue() != "gabriel-greeting-go" ||
				attrs[observability.AttrServiceInstanceID].GetStringValue() != "greeting-test-uid" {
				t.Fatalf("resource attributes missing or mistyped: %+v", attrs)
			}
			if attrs["language"].GetStringValue() != "English" ||
				attrs["name"].GetStringValue() != "Bob" ||
				attrs["greeting"].GetStringValue() != "Hello Bob" ||
				attrs["transport"].GetStringValue() != "unknown" {
				t.Fatalf("unexpected greeting log attributes: %+v", attrs)
			}
			if _, ok := attrs["duration_ns"].GetValue().(*v1.AnyValue_IntValue); !ok {
				t.Fatalf("duration_ns = %+v, want int_value", attrs["duration_ns"])
			}
			if attrs["duration_ns"].GetIntValue() < 0 {
				t.Fatalf("duration_ns = %d, want non-negative", attrs["duration_ns"].GetIntValue())
			}
		}
	}
	if !foundLog {
		t.Fatal("missing canonical greeting log entry")
	}
}

func keyValues(attrs []*v1.KeyValue) map[string]*v1.AnyValue {
	out := make(map[string]*v1.AnyValue, len(attrs))
	for _, attr := range attrs {
		if attr != nil {
			out[attr.GetKey()] = attr.GetValue()
		}
	}
	return out
}
